thriftpy2/protocol/apache_json.py

```diff
@@ -0,0 +1,280 @@
+# -*- coding: utf-8 -*-
+"""
+Transport for json protocol that apache thrift files will understand
+unfortunately, thriftpy2's TJSONProtocol is not compatible with apache's
+"""
+from __future__ import absolute_import
+import json
+import base64
+
+from six import string_types
+from thriftpy2.protocol import TProtocolBase
+from thriftpy2.thrift import TType
+
+
+CTYPES = {
+    TType.BOOL: 'tf',
+    TType.BYTE: 'i8',
+    TType.I16: 'i16',
+    TType.I32: 'i32',
+    TType.I64: 'i64',
+    TType.DOUBLE: 'dbl',
+    TType.STRING: 'str',
+    TType.BINARY: 'str',  # apache sends binary data as base64 encoded
+    TType.STRUCT: 'rec',
+    TType.LIST: 'lst',
+    TType.SET: 'set',
+    TType.MAP: 'map',
+}
+
+JTYPES = {v: k for k, v in CTYPES.items()}
+
+VERSION = 1
+
+
+def flatten(suitable_for_isinstance):
+    """
+    isinstance() can accept a bunch of really annoying different types:
+        * a single type
+        * a tuple of types
+        * an arbitrary nested tree of tuples
+    Return a flattened tuple of the given argument.
+    """
+
+    types = list()
+
+    if not isinstance(suitable_for_isinstance, tuple):
+        suitable_for_isinstance = (suitable_for_isinstance,)
+    for thing in suitable_for_isinstance:
+        if isinstance(thing, tuple):
+            types.extend(flatten(thing))
+        else:
+            types.append(thing)
+    return tuple(types)
+
+
+class TApacheJSONProtocolFactory:
+    @staticmethod
+    def get_protocol(trans):
+        return TApacheJSONProtocol(trans)
+
+
+class TApacheJSONProtocol(TProtocolBase):
+    """
+    Protocol that implements the Apache JSON Protocol
+    """
+
+    def __init__(self, trans):
+        TProtocolBase.__init__(self, trans)
```

Comment:
It's better to call `super(TApacheJSONProtocol, self).__init__(trans)`.

Reply:
I copied the usage from existing protocol classes:

https://github.com/Thriftpy/thriftpy2/blob/17a66b5c3fdef6dd66c5d853c08eafcbffb0b7e4/thriftpy2/protocol/binary.py#L363

https://github.com/Thriftpy/thriftpy2/blob/17a66b5c3fdef6dd66c5d853c08eafcbffb0b7e4/thriftpy2/protocol/compact.py#L127

https://github.com/Thriftpy/thriftpy2/blob/17a66b5c3fdef6dd66c5d853c08eafcbffb0b7e4/thriftpy2/protocol/json.py#L166

Unless you want to change all of them to `super` calls?

Reply:
I think we should not modify the old codes in this PR. Maybe we can change it in the new codes only?

Reply:
@aisk Feel free to create an issue and I will fix it in a 2nd pull request.